src/Client/V2/Access/DatabaseAccess.cs

```diff
@@ -389,11 +411,211 @@ public virtual DatabaseQueryResult Query( string databaseId, DatabaseQuery query
+                            result.Add( orderedIds[i], ScrubDatabaseValue( value ) );
+                            i++;
+                        }
+
+                        yield return result;
+                    }
+                    startingRow += rowsPerCall;
+                    moreToRead = data.HasMoreRows;
+                }
+            }
+            finally
+            {
+                if( info != null )
+                    await StopQueryAsync( databaseId, info.Id, context );
+            }
+        }
+
+        /// <summary>
+        /// Reads all of the values for an async database query and returns them as an async enumerable of dictionaries.
+        /// The keys in the returned dictionary are the field ids, and the values are the values for each field represented
+        /// by the field id. This will throw an <see cref="System.ArgumentException"/> if a field is specified more than
+        /// once in the query due to attempting to add duplicate dictionary keys and QueryValuesAsync should be used instead.
+        /// </summary>
+        /// <param name="databaseId">
+        /// The unique identifier of the EMS database being queried.
+        /// </param>
+        /// <param name="info">
+        /// The asynchronous query info, as retrieved from <see cref="StartQueryAsync(string, DatabaseQuery, CallContext)"/>.
+        /// </param>
+        /// <param name="rowsPerCall">
+        /// The number of rows to read for each API call.
+        /// </param>
+        /// <param name="context">
+        /// The optional call context to include.
+        /// </param>
+        public virtual async IAsyncEnumerable<IReadOnlyDictionary<string, object>> ReadQueryDictionaryAsync( string databaseId, AsyncQueryInfo info, int rowsPerCall = DefaultRowsPerPage, CallContext context = null )
+        {
+            string[] orderedIds = GetOrderedFieldIds( info );
+
+            int startingRow = 0;
+            bool moreToRead = true;
+            while( moreToRead )
+            {
+                AsyncQueryData data = await ReadQueryAsync( databaseId, info.Id, startingRow, startingRow + rowsPerCall - 1, context );
+                foreach( ICollection<object> row in data.Rows )
+                {
+                    int i = 0;
+                    var result = new Dictionary<string, object>();
+                    foreach( object value in row )
+                    {
+                        result.Add( orderedIds[i], ScrubDatabaseValue( value ) );
+                        i++;
+                    }
+
+                    yield return result;
+                }
+                startingRow += rowsPerCall;
+                moreToRead = data.HasMoreRows;
+            }
+        }
+
+        /// <summary>
+        /// Returns the field ids from the query info in order.
+        /// </summary>
+        private string[] GetOrderedFieldIds( AsyncQueryInfo info )
+        {
+            string[] orderedIds = new string[info.Header.Count];
```

Comment:
I am too lazy to lookup the type of info.Header, but maybe we can use the Linq ToArray for this?

Reply:
We'd also need a `Select` here to pull out the id, and I'm not sure what exactly that's going to end up generating, so I still feel a little better about doing it manually even though the linq looks nicer